lib/foreman_tasks/dynflow/daemon.rb

```diff
@@ -1,32 +1,54 @@
 require 'fileutils'
+require 'daemons'
+require 'get_process_mem'
+require 'dynflow/watchers/memory_consumption_watcher'
 
 module ForemanTasks
   class Dynflow::Daemon
+    attr_reader :dynflow_memory_watcher_class,
+                :daemons_class
+
+    # make Daemon dependency injection ready for testing purposes
+    def initialize(
+      dynflow_memory_watcher_class = ::Dynflow::Watchers::MemoryConsumptionWatcher,
+      daemons_class = ::Daemons
+    )
+      @dynflow_memory_watcher_class = dynflow_memory_watcher_class
+      @daemons_class = daemons_class
+    end
+
     # load the Rails environment and initialize the executor
     # in this thread.
-    def run(foreman_root = Dir.pwd)
+    def run(foreman_root = Dir.pwd, options = {})
       STDERR.puts('Starting Rails environment')
       foreman_env_file = File.expand_path('./config/environment.rb', foreman_root)
       unless File.exist?(foreman_env_file)
         raise "#{foreman_root} doesn't seem to be a foreman root directory"
       end
+
+      Foreman::Logging.logger('foreman-tasks').info("Starting dynflow with the following options: #{options}")
+
       ForemanTasks.dynflow.executor!
+
+      if options[:memory_limit] > 0
+        ForemanTasks.dynflow.config.on_init do |world|
+          memory_watcher = initialize_memory_watcher(world, options[:memory_limit], options)
+          world.terminated.on_completion do
+            STDERR.puts("World has been terminated")
+            memory_watcher = nil # the object can be disposed
+          end
+        end
+      end
+
       require foreman_env_file
-      STDERR.puts('Everything ready')
+      STDERR.puts("Everything ready for world: #{(ForemanTasks.dynflow.initialized? ? ForemanTasks.dynflow.world : nil)}")
```

Comment:
`ForemanTasks.dynflow.world.id` is more suitable representation here, then just the `to_s` method.